substrate/frame/vesting/src/lib.rs

```diff
@@ -534,32 +534,25 @@ impl<T: Config> Pallet<T> {
 		if !schedule.is_valid() {
 			return Err(Error::<T>::InvalidScheduleParams.into())
 		};
+
 		let target = T::Lookup::lookup(target)?;
 		let source = T::Lookup::lookup(source)?;
 
-		// Check we can add to this account prior to any storage writes.
-		Self::can_add_vesting_schedule(
-			&target,
-			schedule.locked(),
-			schedule.per_block(),
-			schedule.starting_block(),
-		)?;
-
 		T::Currency::transfer(
 			&source,
 			&target,
 			schedule.locked(),
 			ExistenceRequirement::AllowDeath,
 		)?;
 
-		// We can't let this fail because the currency transfer has already happened.
-		let res = Self::add_vesting_schedule(
+		// If adding this vesting schedule fails, all storage changes are undone due to FRAME's
+		// default transactional layers.
+		Self::add_vesting_schedule(
 			&target,
 			schedule.locked(),
 			schedule.per_block(),
 			schedule.starting_block(),
-		);
-		debug_assert!(res.is_ok(), "Failed to add a schedule when we had to succeed.");
+		)?;
```

Comment:
This means that everyone calling this will need to use `transactional` to revert the changes if there is an error.

Reply:
everyone in frame uses transactional? what am I missing?

All code inside the runtime now assume we can error at any point in the code. when this was written, it wasn't the case, hence the need to "double check"

Reply:
But `VestedTransfer::vested_transfer` function is not a pallet call. If someone implements an error recovery logic it can get messy I suppose.

Maybe it is standard practice now, but I would add a small comment on top of `do_vesting_transfer` that the storage is dirty in case the method becomes public in the future.
And in the doc of `VestedTransfer::vested_transfer` I would also write that the storage can be dirty when error is returned.

Reply:
> But `VestedTransfer::vested_transfer` function is not a pallet call. If someone implements an error recovery logic it can get messy I suppose.

Exactly. If someone handles the error, it will not be transactional. Or you do the transactional handling explicitly in the implementation of the trait.

Reply:
Thanks, I really didn't consider this.

I feel there must be many traits in Polkadot SDK which are vulnerable to the same problem, right?

Reply:
I agree that in general people should probably rarely recover from a dispatch error, and it would be simpler to require the caller to rollback.

Maybe we can make a transition with a new type: `DispatchErrorDirty` or `DispatchErrorToRollback` which means that the storage is dirty and should be rollbacked to handle the error manually.

Or maybe everybody already assume that a `DispatchError` can have a dirty storage.

Reply:
yeah this kind of direction would be nice overall imo